### PR TITLE
Make a full reconstruction and add many new features

### PR DESCRIPTION
**Pull request by Jack-Works**

### CODE REVIEW

**Comment by wfjsw** [comment]
pending review.

File: lib/libbot.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+var Telegram = require('node-telegram-bot-api');
+var cp = require('child_process');
+
+function Bot (config) {
+    var bot = new Telegram(config.bot_api_key, {polling: true});
+    var cli = cp.spawn(config.python3_path, [config.cli_wrapper_path, config.cli_path]);
```
Comment (wfjsw) [inline lib/libbot.js]: consider remove wrapper

File: lib/libnamer.js [hunk 2]
```diff
@@ -1,198 +1,118 @@
...
-    current.unshift(content);
-    while (((gen_title(current)).length * 2) >= 255)
-        current.pop();
-    updateName(gen_title(current));
+function Namer (config, bot, cli) {
+    this.config = config
+    this.prefix = config.prefix
+    this.bot = bot
+    this.cli = cli
+    cli.on('error', process.exit.bind(process));
```
Comment (wfjsw) [inline lib/libnamer.js]: help wanted: work out a genius way to restart cp only.

File: lib/libnamer.js [hunk 3]
```diff
@@ -1,198 +1,118 @@
...
-            if (rmsg)
-                tg.sendMessage(message.chat.id, rmsg);
-        }
+/* doAdd: add a new sub title to the end or the start
+ * (String, TelegramMessage, ?'end') => undefined
+ */
+Namer.prototype.doAdd = function doAdd(content, msg, direction) {
+    var operateRemove = direction == 'end' ? function pop() {this.currentTitle.pop();} : function shift() {this.currentTitle.shift();}
+    var operateAdd = direction == 'end' ? function push(v) {this.currentTitle.push(v);} : function unshift(v) {this.currentTitle.unshift(v);}
+    operateAdd.bind(this)(content.replace(/\s-\s/g, " "));
```
Comment (wfjsw) [inline lib/libnamer.js]: A while should be here too to make sure there is no ' - ' exist after replacing.

File: lib/libbot.js (hunk 1 above)
Comment (Jack-Works) [inline lib/libbot.js]: The wrapper add support for muti-instance support
`argv._.map(function (config){return require(config);}).forEach(Bot);`

File: lib/libnamer.js (hunk 2 above)
Comment (Jack-Works) [inline lib/libnamer.js]: Um, I don't think let the bot run with a previous error is a good idea.
I think we should restart the whole bot.
Why not use forever or systemd to manage the bot status?

File: lib/libnamer.js (hunk 3 above)
Comment (Jack-Works) [inline lib/libnamer.js]: will fix

**Pull request merged**